bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java

```diff
@@ -267,18 +242,40 @@ public void run() {
       }
     }, executorService);
 
+    int timeout = retryOptions.getReadPartialRowTimeoutMillis();
+
+    int streamingBufferSize = retryOptions.getStreamingBufferSize();
+    ResponseQueueReader responseQueueReader =
+        new ResponseQueueReader(timeout, streamingBufferSize, streamingBufferSize, readRowsCall);
+
     StreamingBigtableResultScanner resultScanner =
-        new StreamingBigtableResultScanner(
-          reservedChannel,
-          retryOptions.getStreamingBufferSize(),
-          retryOptions.getReadPartialRowTimeoutMillis(),
-          cancellationToken);
+        new StreamingBigtableResultScanner(reservedChannel, responseQueueReader, cancellationToken);
 
     clientCallService.asyncServerStreamingCall(
         readRowsCall,
         request,
-        new ReadRowsStreamObserver(resultScanner));
+        streamingBufferSize,
+        createClientCallListener(resultScanner));
 
     return resultScanner;
   }
+
+  private ClientCall.Listener<ReadRowsResponse> createClientCallListener(
+      final StreamingBigtableResultScanner resultScanner) {
+    return new ClientCall.Listener<ReadRowsResponse>() {
+      @Override
+      public void onMessage(ReadRowsResponse readRowResponse) {
+        resultScanner.addResult(readRowResponse);
```

Comment:
Not really.  The idea is to maintain this:
  queueSize + outstanding requests > capacity - call request size.

When that is false, request more.

In onMessage, queueSize is incremented by 1 and outstanding requests is decremented by 1 making the sum the same as it was before.

It would seem to me that we should only request more when objects are popped off the queue.